shm/Store.cpp

```diff
@@ -0,0 +1,100 @@
+#include "Store.hpp"
+#include <algorithm>
+#include <iostream>
+#include "Alcohol.hpp"
+#include "Cargo.hpp"
+#include "Fruit.hpp"
+#include "Item.hpp"
+#include "Player.hpp"
+
+Store::Store(Time* time) : time_(time) {
+    GenerateGoodsList();
+}
+
+Cargo* Store::findProduct(Cargo* cargo) {
+    auto it = std::find_if(goodsList_.begin(), goodsList_.end(),
+                           [cargo](const auto& product) { return *cargo == *product; });
+    if (it != goodsList_.end()) {
+        return (*it).get();
+    } else {
+        return nullptr;
+    }
+}
+
+Response Store::buy(Cargo* product, size_t amount, Player* player) {
+    if (product) {
+        auto price = amount * product->getBasePrice();
+        if (player->getAvailableSpace() < amount) {
+            return Response::lack_of_space;
+        }
+        if (product->getAmount() < amount) {
+            return Response::lack_of_cargo;
+        }
+        if (player->getMoney() < price) {
+            return Response::lack_of_money;
+        }
+
+        auto cargo = product->clone(amount);
+        time_->AddObserver(cargo.get());
```

Comment:
Hmm this is tricky. Cargo should register obserwer and delete it. Store class should don't know about specific implememntation of Cargo.